modulus/models/unet/unet3d.py

```diff
@@ -0,0 +1,510 @@
+            x = layer(x)
+        return x
+
+
+class DecoderBlock(nn.Module):
+    """
+    A decoder block that sequentially applies multiple transposed convolutional blocks, optionally concatenating features from the corresponding encoder.
+
+    Parameters:
+    ----------
+        in_channels (int): Number of channels in the input.
+        feature_map_channels (List[int]): List of the number of channels for each deconv block within this decoder.
+        model_depth (int): Number of times the deconv operation should be repeated.
+        num_conv_blocks (int): Number of deconvolutional blocks per depth level.
+
+    Returns:
+    -------
+        torch.Tensor: The processed output tensor.
+    """
+
+    def __init__(
+        self,
+        out_channels: int,
+        feature_map_channels: List[int],
+        model_depth: int = 3,
+        num_conv_blocks: int = 2,
+    ):
+        super().__init__()
+
+        if len(feature_map_channels) != model_depth * num_conv_blocks + 1:
+            raise ValueError(
+                "The length of feature_map_channels in the decoder block should be equal to model_depth * num_conv_blocks + 1"
+            )
+
+        self.layers = nn.ModuleList()
+        current_channels = feature_map_channels[0]
+        feature_map_channels = feature_map_channels[1:]
+
+        for depth in range(model_depth):
+            for i in range(num_conv_blocks):
+                if i == 0:
+                    self.layers.append(
+                        ConvTranspose(
+                            in_channels=current_channels,
+                            out_channels=current_channels,
+                        )
+                    )
+                    current_channels += feature_map_channels[
+                        depth * num_conv_blocks + i
+                    ]
+
+                self.layers.append(
+                    ConvBlock(
+                        in_channels=current_channels,
+                        out_channels=feature_map_channels[depth * num_conv_blocks + i],
+                    )
+                )
+                current_channels = feature_map_channels[depth * num_conv_blocks + i]
+
+        # Final convolution
+        self.layers.append(
+            ConvBlock(
+                in_channels=current_channels,
+                out_channels=out_channels,
+                activation=None,
+                normalization=None,
+            )
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        for layer in self.layers:
+            x = layer(x)
+        return x
+
+
+@dataclass
+class MetaData(ModelMetaData):
+    name: str = "UNet3D"
+    # Optimization
+    jit: bool = False
+    cuda_graphs: bool = True
+    amp: bool = True
+    # Inference
+    onnx_cpu: bool = True
+    onnx_gpu: bool = True
+    onnx_runtime: bool = True
+    # Physics informed
+    var_dim: int = 1
+    func_torch: bool = False
+    auto_grad: bool = False
+
+
+class UNet3D(Module):
```

Comment:
Linking this issue here: https://github.com/NVIDIA/modulus/issues/512 - we should try to add support for 2D soon. In that case, should the name be `UNet` ?

Reply:
Renamed to UNet. Seems like we have a volunteer to add support for 2D after this PR is merged: https://github.com/NVIDIA/modulus/issues/512#issuecomment-2138534646